File: server/src/controllers/authController.ts
import { Request, Response } from 'express';
import { query } from '../config/db.js';
import bcrypt from 'bcryptjs';
import { v4 as uuidv4 } from 'uuid';
import { toMySQLDateTime } from '../utils/dataHelpers.js';
import jwt from 'jsonwebtoken';

const JWT_SECRET = process.env.JWT_SECRET || 'constructos-secret-key-change-in-production';

// Register new user
export const register = async (req: Request, res: Response) => {
  try {
    const { name, email, phone, password, role } = req.body;
    
    // Validation
    if (!name || !email || !password) {
      return res.status(400).json({ error: 'Vui lòng nhập đầy đủ tên, email và mật khẩu' });
    }
    
    if (password.length < 6) {
      return res.status(400).json({ error: 'Mật khẩu phải có ít nhất 6 ký tự' });
    }
    
    // Check if user already exists
    const existing = await query<any[]>(
      'SELECT id FROM users WHERE email = ?',
      [email]
    );
    
    if (existing.length > 0) {
      return res.status(400).json({ error: 'Email đã tồn tại' });
    }
    
    // Hash password
    const saltRounds = 10;
    const passwordHash = await bcrypt.hash(password, saltRounds);
    
    // Validate and resolve role_id
    // Frontend may send role name (string) or role UUID
    let roleId = role;
    if (!roleId) {
      // Try to get a default role (first role in database)
      try {
        const defaultRole = await query<any[]>(
          'SELECT id FROM roles LIMIT 1'
        );
        if (defaultRole.length > 0) {
          roleId = defaultRole[0].id;
        } else {
          return res.status(400).json({ error: 'No roles found in database. Please create at least one role first.' });
        }
      } catch (roleError) {
        return res.status(400).json({ error: 'Role (role_id) is required and no default role found' });
      }
    } else {
      // Check if role is a UUID format (36 chars with dashes)
      const isUUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(roleId);
      
      if (!isUUID) {
        // Role is a name string, find the UUID by name
        try {
          // Map common role names to database role names
          const roleNameMap: { [key: string]: string } = {
            'admin': 'admin',
            'project_manager': 'project_manager',
            'accountant': 'accountant',
            'warehouse': 'warehouse',
            'site_manager': 'site_manager',
            'engineer': 'engineer',
            'client': 'client',
            'kế toán': 'accountant',
            'quản lý công trường': 'site_manager',
            'quản lý dự án': 'project_manager',
          };
          
          const mappedRoleName = roleNameMap[roleId.toLowerCase()] || roleId.toLowerCase();
          const roleResult = await query<any[]>(
            'SELECT id FROM roles WHERE LOWER(name) = $1',
            [mappedRoleName]
          );
          
          if (roleResult.length === 0) {
            return res.status(400).json({ error: `Role "${roleId}" not found. Please use a valid role name or UUID.` });
          }
          
          roleId = roleResult[0].id;
        } catch (roleError: any) {
          return res.status(400).json({ error: `Invalid role: ${roleError.message}` });
        }
      } else {
        // Role is a UUID, validate it exists
        try {
          const roleCheck = await query<any[]>(
            'SELECT id FROM roles WHERE id = $1',
            [roleId]
          );
          if (roleCheck.length === 0) {
            return res.status(400).json({ error: `Role with id ${roleId} does not exist` });
          }
        } catch (roleError: any) {
          return res.status(400).json({ error: `Invalid role_id: ${roleError.message}` });
        }
      }
    }
    
    // Create user
    const id = uuidv4();
    const createdAt = toMySQLDateTime();
    
    await query(
      `INSERT INTO users (
        id, name, email, phone, password_hash, role, status, created_at, updated_at
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        id,
        name,
        email,
        phone || '',
        passwordHash,
        roleId,
        'active',
        createdAt,
        createdAt,
      ]
    );
    
    // Get created user (without password)
    const newUser = await query<any[]>(
      'SELECT id, name, email, phone, role, status, avatar, created_at, updated_at FROM users WHERE id = ?',
      [id]
    );
    
    // Generate JWT token
    const token = jwt.sign(
      { userId: id, email, role: role || 'client' },
      JWT_SECRET,
      { expiresIn: '7d' }
    );
    
    res.status(201).json({
      user: newUser[0],
      token,
    });
  } catch (error: any) {
    console.error('Error registering user:', error);
    res.status(500).json({ 
      error: 'Đăng ký thất bại',
      message: error.message 
    });
  }
};

// Login
export const login = async (req: Request, res: Response) => {
  try {
    const { email, password } = req.body;
    
    // Validation
    if (!email || !password) {
      return res.status(400).json({ error: 'Vui lòng nhập email và mật khẩu' });
    }
    
    // Find user
    const users = await query<any[]>(
      'SELECT * FROM users WHERE email = ?',
      [email]
    );
    
    if (users.length === 0) {
      return res.status(401).json({ error: 'Email hoặc mật khẩu không đúng' });
    }
    
    const user = users[0];
    
    // Check if user is active
    if (user.status !== 'active') {
      return res.status(403).json({ error: 'Tài khoản chưa được kích hoạt' });
    }
    
    // Verify password
    const isValidPassword = await bcrypt.compare(password, user.password_hash);
    
    if (!isValidPassword) {
      return res.status(401).json({ error: 'Email hoặc mật khẩu không đúng' });
    }
    
    // Generate JWT token
    const token = jwt.sign(
      { userId: user.id, email: user.email, role: user.role },
      JWT_SECRET,
      { expiresIn: '7d' }
    );
    
    // Return user (without password)
    res.json({
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        phone: user.phone,
        role: user.role,
        status: user.status,
        avatar: user.avatar,
        createdAt: user.created_at,
        updatedAt: user.updated_at,
      },
      token,
    });
  } catch (error: any) {
    console.error('Error logging in:', error);
    res.status(500).json({ 
      error: 'Đăng nhập thất bại',
      message: error.message 
    });
  }
};

// Forgot password - Generate reset token
export const forgotPassword = async (req: Request, res: Response) => {
  try {
    const { email } = req.body;
    
    if (!email) {
      return res.status(400).json({ error: 'Vui lòng nhập email' });
    }
    
    // Find user
    const users = await query<any[]>(
      'SELECT id, name, email FROM users WHERE email = ?',
      [email]
    );
    
    // Always return success (security: don't reveal if email exists)
    if (users.length > 0) {
      const user = users[0];
      
      // Generate reset token
      const resetToken = jwt.sign(
        { userId: user.id, email: user.email, type: 'password_reset' },
        JWT_SECRET,
        { expiresIn: '1h' }
      );
      
      // In production, send email with reset link
      // For now, we'll just return the token (in production, send via email)
      console.log(`Password reset token for ${email}: ${resetToken}`);
      
      // TODO: Send email with reset link
      // await sendPasswordResetEmail(user.email, resetToken);
    }
    
    // Always return success message
    res.json({ 
      message: 'Nếu email tồn tại, liên kết đặt lại mật khẩu đã được gửi' 
    });
  } catch (error: any) {
    console.error('Error in forgot password:', error);
    res.status(500).json({ 
      error: 'Xử lý yêu cầu đặt lại mật khẩu thất bại',
      message: error.message 
    });
  }
};

// Reset password with token
export const resetPassword = async (req: Request, res: Response) => {
  try {
    const { token, newPassword } = req.body;
    
    if (!token || !newPassword) {
      return res.status(400).json({ error: 'Vui lòng nhập token và mật khẩu mới' });
    }
    
    if (newPassword.length < 6) {
      return res.status(400).json({ error: 'Mật khẩu phải có ít nhất 6 ký tự' });
    }
    
    // Verify token
    let decoded: any;
    try {
      decoded = jwt.verify(token, JWT_SECRET);
      
      if (decoded.type !== 'password_reset') {
        return res.status(400).json({ error: 'Loại token không hợp lệ' });
      }
    } catch (error) {
      return res.status(400).json({ error: 'Token không hợp lệ hoặc đã hết hạn' });
    }
    
    // Hash new password
    const saltRounds = 10;
    const passwordHash = await bcrypt.hash(newPassword, saltRounds);
    
    // Update password
    const updatedAt = toMySQLDateTime();
    await query(
      'UPDATE users SET password_hash = ?, updated_at = ? WHERE id = ?',
      [passwordHash, updatedAt, decoded.userId]
    );
    
    res.json({ message: 'Đặt lại mật khẩu thành công' });
  } catch (error: any) {
    console.error('Error resetting password:', error);
    res.status(500).json({ 
      error: 'Đặt lại mật khẩu thất bại',
      message: error.message 
    });
  }
};

// Verify token (for checking if reset token is valid)
export const verifyToken = async (req: Request, res: Response) => {
  try {
    const { token } = req.body;
    
    if (!token) {
      return res.status(400).json({ error: 'Vui lòng nhập token' });
    }
    
    try {
      const decoded = jwt.verify(token, JWT_SECRET) as any;
      
      if (decoded.type === 'password_reset') {
        res.json({ valid: true, type: 'password_reset' });
      } else {
        res.json({ valid: true, type: 'auth' });
      }
    } catch (error) {
      res.json({ valid: false });
    }
  } catch (error: any) {
    console.error('Error verifying token:', error);
    res.status(500).json({ 
      error: 'Xác thực token thất bại',
      message: error.message 
    });
  }
};

File: server.net/Controllers/AuthController.cs
using ConstructOs.Server.Config;
using ConstructOs.Server.Utils;
using Microsoft.AspNetCore.Mvc;
using BCrypt.Net;
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Microsoft.IdentityModel.Tokens;
using Dapper;

namespace ConstructOs.Server.Controllers;

[ApiController]
[Route("api/auth")]
public class AuthController : ControllerBase
{
    private readonly IDatabaseService _db;
    private readonly IConfiguration _configuration;
    private readonly ILogger<AuthController> _logger;

    public AuthController(IDatabaseService db, IConfiguration configuration, ILogger<AuthController> logger)
    {
        _db = db;
        _configuration = configuration;
        _logger = logger;
    }

    private string GenerateJwtToken(string userId, string email, string role)
    {
        var jwtSecret = _configuration["JWT_SECRET"] ?? "constructos-secret-key-change-in-production";
        var key = Encoding.UTF8.GetBytes(jwtSecret);

        var claims = new[]
        {
            new Claim("userId", userId),
            new Claim(ClaimTypes.Email, email),
            new Claim(ClaimTypes.Role, role)
        };

        var tokenDescriptor = new SecurityTokenDescriptor
        {
            Subject = new ClaimsIdentity(claims),
            Expires = DateTime.UtcNow.AddDays(7),
            SigningCredentials = new SigningCredentials(
                new SymmetricSecurityKey(key),
                SecurityAlgorithms.HmacSha256Signature)
        };

        var tokenHandler = new JwtSecurityTokenHandler();
        var token = tokenHandler.CreateToken(tokenDescriptor);
        return tokenHandler.WriteToken(token);
    }

    [HttpPost("register")]
    public async Task<IActionResult> Register([FromBody] RegisterRequest request)
    {
        try
        {
            if (string.IsNullOrEmpty(request.Name) || 
                string.IsNullOrEmpty(request.Email) || 
                string.IsNullOrEmpty(request.Password))
            {
                return BadRequest(new { error = "Vui lòng nhập đầy đủ tên, email và mật khẩu" });
            }

            if (request.Password.Length < 6)
            {
                return BadRequest(new { error = "Mật khẩu phải có ít nhất 6 ký tự" });
            }

            // Check if user already exists
            var existing = await _db.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT id FROM users WHERE email = ?",
                new { request.Email });

            if (existing != null)
            {
                return BadRequest(new { error = "Email đã tồn tại" });
            }

            // Hash password
            var passwordHash = BCrypt.Net.BCrypt.HashPassword(request.Password, 10);

            // Validate and resolve role_id
            string? roleId = request.Role;
            if (string.IsNullOrEmpty(roleId))
            {
                // Try to get default role "construction_department"
                var defaultRole = await _db.QueryFirstOrDefaultAsync<dynamic>(
                    "SELECT id FROM roles WHERE LOWER(name) = ? OR LOWER(name) = ?",
                    new { roleName1 = "construction_department", roleName2 = "phòng xây dựng" });

                if (defaultRole != null)
                {
                    roleId = defaultRole.id?.ToString();
                }
                else
                {
                    // Fallback: get first role
                    var fallbackRole = await _db.QueryFirstOrDefaultAsync<dynamic>(
                        "SELECT id FROM roles LIMIT 1");
                    
                    if (fallbackRole != null)
                    {
                        roleId = fallbackRole.id?.ToString();
                        _logger.LogWarning("Role 'construction_department' not found, using first available role");
                    }
                    else
                    {
                        return BadRequest(new { error = "No roles found in database. Please create at least one role first." });
                    }
                }
            }
            else
            {
                // Check if role is a UUID format
                var isUUID = System.Text.RegularExpressions.Regex.IsMatch(roleId, 
                    @"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$", 
                    System.Text.RegularExpressions.RegexOptions.IgnoreCase);

                if (!isUUID)
                {
                    // Role is a name string, find the UUID by name
                    var roleNameMap = new Dictionary<string, string>
                    {
                        { "admin", "admin" },
                        { "project_manager", "project_manager" },
                        { "accountant", "accountant" },
                        { "warehouse", "warehouse" },
                        { "site_manager", "site_manager" },
                        { "engineer", "engineer" },
                        { "client", "client" },
                        { "kế toán", "accountant" },
                        { "quản lý công trường", "site_manager" },
                        { "quản lý dự án", "project_manager" }
                    };

                    var mappedRoleName = roleNameMap.ContainsKey(roleId.ToLower()) 
                        ? roleNameMap[roleId.ToLower()] 
                        : roleId.ToLower();

                    var roleResult = await _db.QueryFirstOrDefaultAsync<dynamic>(
                        "SELECT id FROM roles WHERE LOWER(name) = ?",
                        new { roleName = mappedRoleName });

                    if (roleResult == null)
                    {
                        return BadRequest(new { error = $"Role \"{roleId}\" not found. Please use a valid role name or UUID." });
                    }

                    roleId = roleResult.id?.ToString();
                }
                else
                {
                    // Role is a UUID, validate it exists
                    var roleCheck = await _db.QueryFirstOrDefaultAsync<dynamic>(
                        "SELECT id FROM roles WHERE id = ?",
                        new { roleId });

                    if (roleCheck == null)
                    {
                        return BadRequest(new { error = $"Role with id {roleId} does not exist" });
                    }
                }
            }

            // Create user
            var id = Guid.NewGuid().ToString();
            var createdAt = DataHelpers.ToMySQLDateTime();

            await _db.ExecuteAsync(
                @"INSERT INTO users (
                    id, name, email, phone, password_hash, role, status, created_at, updated_at
                ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
                new
                {
                    id,
                    name = request.Name,
                    email = request.Email,
                    phone = request.Phone ?? "",
                    passwordHash,
                    role = roleId,
                    status = "active",
                    createdAt,
                    updatedAt = createdAt
                });

            // Get created user (without password)
            var newUser = await _db.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT id, name, email, phone, role, status, avatar, created_at, updated_at FROM users WHERE id = ?",
                new { id });

            // Generate JWT token
            var token = GenerateJwtToken(id, request.Email, request.Role ?? "client");

            return StatusCode(201, new
            {
                user = newUser,
                token
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error registering user");
            return StatusCode(500, new
            {
                error = "Đăng ký thất bại",
                message = ex.Message
            });
        }
    }

    [HttpPost("login")]
    public async Task<IActionResult> Login([FromBody] LoginRequest request)
    {
        try
        {
            if (string.IsNullOrEmpty(request.Email) || string.IsNullOrEmpty(request.Password))
            {
                return BadRequest(new { error = "Vui lòng nhập email và mật khẩu" });
            }

            // Find user
            var user = await _db.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT * FROM users WHERE email = ?",
                new { request.Email });

            if (user == null)
            {
                return Unauthorized(new { error = "Email hoặc mật khẩu không đúng" });
            }

            // Check if user is active
            if (user.status?.ToString() != "active")
            {
                return StatusCode(403, new { error = "Tài khoản chưa được kích hoạt" });
            }

            // Verify password
            var passwordHash = user.password_hash?.ToString() ?? "";
            var isValidPassword = BCrypt.Net.BCrypt.Verify(request.Password, passwordHash);

            if (!isValidPassword)
            {
                return Unauthorized(new { error = "Email hoặc mật khẩu không đúng" });
            }

            // Generate JWT token
            var role = user.role?.ToString() ?? "client";
            var token = GenerateJwtToken(user.id?.ToString() ?? "", user.email?.ToString() ?? "", role);

            // Return user (without password)
            return Ok(new
            {
                user = new
                {
                    id = user.id,
                    name = user.name,
                    email = user.email,
                    phone = user.phone,
                    role = user.role,
                    status = user.status,
                    avatar = user.avatar,
                    createdAt = user.created_at,
                    updatedAt = user.updated_at
                },
                token
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error logging in");
            return StatusCode(500, new
            {
                error = "Đăng nhập thất bại",
                message = ex.Message
            });
        }
    }

    [HttpPost("forgot-password")]
    public async Task<IActionResult> ForgotPassword([FromBody] ForgotPasswordRequest request)
    {
        try
        {
            if (string.IsNullOrEmpty(request.Email))
            {
                return BadRequest(new { error = "Vui lòng nhập email" });
            }

            // Find user
            var user = await _db.QueryFirstOrDefaultAsync<dynamic>(
                "SELECT id, name, email FROM users WHERE email = ?",
                new { request.Email });

            // Always return success (security: don't reveal if email exists)
            if (user != null)
            {
                // Generate reset token
                var resetToken = GenerateJwtToken(user.id?.ToString() ?? "", user.email?.ToString() ?? "", "password_reset");
                
                // In production, send email with reset link
                _logger.LogInformation($"Password reset token for {request.Email}: {resetToken}");
                
                // TODO: Send email with reset link
            }

            // Always return success message
            return Ok(new
            {
                message = "Nếu email tồn tại, liên kết đặt lại mật khẩu đã được gửi"
            });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error in forgot password");
            return StatusCode(500, new
            {
                error = "Xử lý yêu cầu đặt lại mật khẩu thất bại",
                message = ex.Message
            });
        }
    }

    [HttpPost("reset-password")]
    public async Task<IActionResult> ResetPassword([FromBody] ResetPasswordRequest request)
    {
        try
        {
            if (string.IsNullOrEmpty(request.Token) || string.IsNullOrEmpty(request.NewPassword))
            {
                return BadRequest(new { error = "Vui lòng nhập token và mật khẩu mới" });
            }

            if (request.NewPassword.Length < 6)
            {
                return BadRequest(new { error = "Mật khẩu phải có ít nhất 6 ký tự" });
            }

            // Verify token
            var jwtSecret = _configuration["JWT_SECRET"] ?? "constructos-secret-key-change-in-production";
            var key = Encoding.UTF8.GetBytes(jwtSecret);
            var tokenHandler = new JwtSecurityTokenHandler();

            SecurityToken? validatedToken;
            ClaimsPrincipal? principal;
            try
            {
                principal = tokenHandler.ValidateToken(request.Token, new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = new SymmetricSecurityKey(key),
                    ValidateIssuer = false,
                    ValidateAudience = false,
                    ValidateLifetime = true,
                    ClockSkew = TimeSpan.Zero
                }, out validatedToken);
            }
            catch
            {
                return BadRequest(new { error = "Token không hợp lệ hoặc đã hết hạn" });
            }

            var userId = principal?.FindFirst("userId")?.Value;
            if (string.IsNullOrEmpty(userId))
            {
                return BadRequest(new { error = "Token không hợp lệ" });
            }

            // Hash new password
            var passwordHash = BCrypt.Net.BCrypt.HashPassword(request.NewPassword, 10);

            // Update password
            var updatedAt = DataHelpers.ToMySQLDateTime();
            await _db.ExecuteAsync(
                "UPDATE users SET password_hash = ?, updated_at = ? WHERE id = ?",
                new { passwordHash, updatedAt, userId });

            return Ok(new { message = "Đặt lại mật khẩu thành công" });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error resetting password");
            return StatusCode(500, new
            {
                error = "Đặt lại mật khẩu thất bại",
                message = ex.Message
            });
        }
    }

    [HttpPost("verify-token")]
    public async Task<IActionResult> VerifyToken([FromBody] VerifyTokenRequest request)
    {
        try
        {
            if (string.IsNullOrEmpty(request.Token))
            {
                return BadRequest(new { error = "Vui lòng nhập token" });
            }

            var jwtSecret = _configuration["JWT_SECRET"] ?? "constructos-secret-key-change-in-production";
            var key = Encoding.UTF8.GetBytes(jwtSecret);
            var tokenHandler = new JwtSecurityTokenHandler();

            try
            {
                var principal = tokenHandler.ValidateToken(request.Token, new TokenValidationParameters
                {
                    ValidateIssuerSigningKey = true,
                    IssuerSigningKey = new SymmetricSecurityKey(key),
                    ValidateIssuer = false,
                    ValidateAudience = false,
                    ValidateLifetime = true,
                    ClockSkew = TimeSpan.Zero
                }, out _);

                var type = principal?.FindFirst(ClaimTypes.Role)?.Value == "password_reset" 
                    ? "password_reset" 
                    : "auth";

                return Ok(new { valid = true, type });
            }
            catch
            {
                return Ok(new { valid = false });
            }
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error verifying token");
            return StatusCode(500, new
            {
                error = "Xác thực token thất bại",
                message = ex.Message
            });
        }
    }
}

public class RegisterRequest
{
    public string Name { get; set; } = string.Empty;
    public string Email { get; set; } = string.Empty;
    public string? Phone { get; set; }
    public string Password { get; set; } = string.Empty;
    public string? Role { get; set; }
}

public class LoginRequest
{
    public string Email { get; set; } = string.Empty;
    public string Password { get; set; } = string.Empty;
}

public class ForgotPasswordRequest
{
    public string Email { get; set; } = string.Empty;
}

public class ResetPasswordRequest
{
    public string Token { get; set; } = string.Empty;
    public string NewPassword { get; set; } = string.Empty;
}

public class VerifyTokenRequest
{
    public string Token { get; set; } = string.Empty;
}

